File: go/SlidingWindow/121_Best_Time_to_Buy_and_Sell_Stock/121_Best_Time_to_Buy_and_Sell_Stock.go
package besttimetobuyandsellstock

func maxProfit(prices []int) int {
	var maxProfit int
	leftPointer := 0
	rightPointer := 1

	for rightPointer < len(prices) {
		if prices[leftPointer] < prices[rightPointer] {
			profit := prices[rightPointer] - prices[leftPointer]
			maxProfit = max(maxProfit, profit)
		} else {
			leftPointer = rightPointer
		}
		rightPointer++
	}

	return maxProfit
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

File: kotlin/121-Best-Time-To-Buy-and-Sell-Stock.kt
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
class Solution {
    fun maxProfit(prices: IntArray): Int {
        var maxProfit:Int = 0; var buy:Int = prices[0]
        for (sell in prices) {
            if (sell < buy) { buy = sell }
            val profit = sell - buy
            maxProfit = maxOf(maxProfit, profit)
        }
        return maxProfit
    }
}
